staging/src/k8s.io/component-base/metrics/legacyregistry/registry.go

```diff
@@ -17,112 +17,53 @@ limitations under the License.
 package legacyregistry
 
 import (
-	"fmt"
-	"net/http"
-	"reflect"
-	"sync"
-
 	"github.com/prometheus/client_golang/prometheus"
 	"github.com/prometheus/client_golang/prometheus/promhttp"
-	dto "github.com/prometheus/client_model/go"
-
-	apimachineryversion "k8s.io/apimachinery/pkg/version"
 	"k8s.io/component-base/metrics"
+	"net/http"
 )
 
 var (
-	globalRegistryFactory = metricsRegistryFactory{
-		registerQueue:     make([]metrics.Registerable, 0),
-		mustRegisterQueue: make([]metrics.Registerable, 0),
-		globalRegistry:    noopRegistry{},
-	}
+	defaultRegistry = metrics.NewKubeRegistry()
 )
 
-type noopRegistry struct{}
-
-func (noopRegistry) Register(metrics.Registerable) error     { return nil }
-func (noopRegistry) MustRegister(...metrics.Registerable)    {}
-func (noopRegistry) RawRegister(prometheus.Collector) error  { return nil }
-func (noopRegistry) RawMustRegister(...prometheus.Collector) {}
-func (noopRegistry) Unregister(metrics.Registerable) bool    { return true }
-func (noopRegistry) Gather() ([]*dto.MetricFamily, error)    { return nil, nil }
-
-type metricsRegistryFactory struct {
-	globalRegistry    metrics.KubeRegistry
-	kubeVersion       *apimachineryversion.Info
-	registrationLock  sync.Mutex
-	registerQueue     []metrics.Registerable
-	mustRegisterQueue []metrics.Registerable
+func init() {
+	RawMustRegister(prometheus.NewProcessCollector(prometheus.ProcessCollectorOpts{}))
+	RawMustRegister(prometheus.NewGoCollector())
 }
 
-// HandlerForGlobalRegistry returns a http handler for the global registry. This
-// allows us to return a handler for the global registry without having to expose
-// the global registry itself directly.
-func HandlerForGlobalRegistry(opts promhttp.HandlerOpts) http.Handler {
-	return promhttp.HandlerFor(globalRegistryFactory.globalRegistry, opts)
+// Handler returns an HTTP handler for the DefaultGatherer. It is
+// already instrumented with InstrumentHandler (using "prometheus" as handler
+// name).
+//
+// Deprecated: Please note the issues described in the doc comment of
+// InstrumentHandler. You might want to consider using promhttp.Handler instead.
+func Handler() http.Handler {
```

Comment:
if it is deprecated, why do we change the signature?

Reply:
This method intends to replace the method calls to the original prometheus global Handler (https://github.com/prometheus/client_golang/blob/7ed96b33bd86448bfd5915bf4f0f96b7eb826108/prometheus/http.go#L47-L55), which has the same method signature as this method and is also similarly deprecated (but in use in our codebase).

Reply:
Specifically, we require this for backwards compatibility reasons. The InstrumentedHandler effectively adds a series of summary metrics which are, by default, added to the global prometheus registry. This method allows us to maintain parity with the metrics that the native global prometheus registry.